File: 0396-rotate-function/0396-rotate-function.js
var maxRotateFunction = function(nums) {
    const n = nums.length;
    
    let totalSum = 0;
    let F = 0;
    
    for (let i = 0; i < n; i++) {
        totalSum += nums[i];
        F += i * nums[i];
    }
    
    let maxValue = F;
    
    for (let k = 1; k < n; k++) {
        F = F + totalSum - n * nums[n - k];
        maxValue = Math.max(maxValue, F);
    }
    
    return maxValue;
};

File: 0396-rotate-function/0396-rotate-function.py
class Solution:
    def maxRotateFunction(self, nums):
        n = len(nums)
        total_sum = sum(nums)
        
        # F(0)
        F = sum(i * num for i, num in enumerate(nums))
        max_value = F
        
        # Recurrence
        for k in range(1, n):
            F = F + total_sum - n * nums[-k]
            max_value = max(max_value, F)
        
        return max_value